src/tracing.rs

```diff
@@ -0,0 +1,366 @@
+use crate::{
+    callframe::{Callframe, NearCallFrame},
+    decommit::is_kernel,
+    predication::{self, Predicate},
+    VirtualMachine,
+};
+use eravm_stable_interface::*;
+use std::cmp::Ordering;
+
+impl<T> StateInterface for VirtualMachine<T> {
+    fn read_register(&self, register: u8) -> (u256::U256, bool) {
+        (
+            self.state.registers[register as usize],
+            self.state.register_pointer_flags & (1 << register) != 0,
+        )
+    }
+
+    fn set_register(&mut self, register: u8, value: u256::U256, is_pointer: bool) {
+        self.state.registers[register as usize] = value;
+
+        self.state.register_pointer_flags &= !(1 << register);
+        self.state.register_pointer_flags |= u16::from(is_pointer) << register;
+    }
+
+    fn number_of_callframes(&self) -> usize {
+        self.state
+            .previous_frames
+            .iter()
+            .map(|frame| frame.near_calls.len() + 1)
+            .sum::<usize>()
+            + self.state.current_frame.near_calls.len()
+            + 1
+    }
+
+    fn callframe(&mut self, n: usize) -> impl CallframeInterface + '_ {
+        for far_frame in std::iter::once(&mut self.state.current_frame)
+            .chain(self.state.previous_frames.iter_mut())
```

Comment:
Dumb question: what is the iteration order here? AFAIU, `previous_frames` have stack ordering (the most recent frame is last), but the current frame is prepended, not appended.

Reply:
Looks like I need to add some .rev